ElectricCar.cpp

```diff
@@ -2,47 +2,15 @@
 #include <iostream>
 
 ElectricCar::ElectricCar(std::unique_ptr<IEngine> engine)
-    : engine_(std::move(engine)) {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-}
+    : engine_(std::move(engine)) {}
 
-ElectricCar::~ElectricCar() {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-}
-void ElectricCar::turnLeft() const {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-}
-void ElectricCar::turnRight() const {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-}
-void ElectricCar::brake() const {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-}
-void ElectricCar::accelerate(Speed speed) const {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-}
-void ElectricCar::start_engine() const {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-  engine_->start();
-}
-void ElectricCar::stop_engine() const {
-  std::cout << __FUNCTION__ << ": " << __FILE__ << ": " << __LINE__
-            << std::endl;
-  engine_->stop();
-}
+ElectricCar::~ElectricCar() {}
+void ElectricCar::turnLeft() const {}
+void ElectricCar::turnRight() const {}
+void ElectricCar::brake() const {}
+void ElectricCar::accelerate(Speed speed) const {}
+void ElectricCar::start_engine() const { engine_->start(); }
+void ElectricCar::stop_engine() const { engine_->stop(); }
```

Comment:
that's also something we do not want. You are overriding those methods with empty implementations. You should use your base class implementation, so do not override them. Simply remove them from cpp and hpp. The same for other types of cars.